ebean-migration/src/main/java/io/ebean/migration/MigrationRunner.java

```diff
@@ -22,41 +22,40 @@
 
   private final MigrationConfig migrationConfig;
 
-  private List<LocalMigrationResource> checkMigrations;
+  private final List<LocalMigrationResource> checkMigrations;
 
   public MigrationRunner(MigrationConfig migrationConfig) {
     this.migrationConfig = migrationConfig;
+    this.checkMigrations = new ArrayList<>();
   }
 
   /**
-   * Return the migrations that would be applied if the migration is run.
+   * Return the migrations that would be applied if the migration is invoke.
    */
   public List<LocalMigrationResource> checkState() {
-    run(migrationConfig.createConnection(), true);
-    return checkMigrations;
+    return checkState(getMigrationConfig().createConnection());
   }
 
   /**
-   * Return the migrations that would be applied if the migration is run.
+   * Return the migrations that would be applied if the migration is invoke.
    */
   public List<LocalMigrationResource> checkState(DataSource dataSource) {
-    run(getConnection(dataSource), true);
-    return checkMigrations;
+    return checkState(getConnection(dataSource));
   }
 
   /**
-   * Return the migrations that would be applied if the migration is run.
+   * Return the migrations that would be applied if the migration is invoke.
    */
   public List<LocalMigrationResource> checkState(Connection connection) {
-    run(connection, true);
-    return checkMigrations;
+    invoke(connection, true);
+    return getCheckMigrations();
   }
 
   /**
-   * Run by creating a DB connection from driver, url, username defined in MigrationConfig.
+   * Run by creating a DB connection from driver, url, username defined in getMigrationConfig().
    */
   public void run() {
-    run(migrationConfig.createConnection());
+    run(getMigrationConfig().createConnection());
```

Comment:
Why is getMigrationConfig() here given migrationConfig is a private final ?

Reply:
Problem is that MigrationRunner creates the database tables even if migration fails. As a result of this migration cannot not be applied after first creation fails.

The above code uses the getter newly added to the code to be able to override certain setting from extended code.

Reply:
One more comment on private final migrationConfig. MigrationConfig is private final but it is not immutable. Thus, some settings can be modified even if migrationConfig is final.

Reply:
Hi Rob,

Sorry, to bother you again. I'd like to give some more details on the pull request.

The reason to use a getter for migrationConfig is to be able to override the protected invoke method. 

BTW: I renamed it to better differentiate between the migration _run_ method and the checkState internally used private _run_ method. In terms of being able to extend MigrationRunner another naming for the internal private _run_ method makes sense from my point-of-view.

Regarding the getter pattern:

Please refer to [https://stackoverflow.com/questions/8466790/java-conventions-use-getters-setters-within-the-class](https://stackoverflow.com/questions/8466790/java-conventions-use-getters-setters-within-the-class)

Let's assume you would like to modify the invoke method and migrationConfig would not have a getter method and you only pass the migrationConfig as constructor arg to the super class you wouldn't be able to access the private field migrationConfig from inside the extended invoke method later on. Hence, extending the class is nearly impossible without copying everything.

If the original MigrationRunner would only get used of the provided constructor arg migrationConfig, extending would also be limited. That's why I moved from field access to getter.

Hope, it is a little clearer now what the approach is.

Regards.. :-)

Reply:
> Regarding the getter pattern:

Well, the [ using a member variable means "no surprises here" ]comment is also the point I'd make but I'd be stronger and say that if there is a need to use getters internally then that actually suggests a design issue/fault to me.  That is, there isn't any good need for a getter method to be used internally in a class unless there is something wrong with the design (or it really is not a getter but computes something).

That is, the field could just be made _protected_ and hence available to a class/code extending it. Changing internal code to use getters shouts to me that there is a design issue. Why didn't you look to change the field to protected?

 
> MigrationRunner creates the database tables even if migration fails. 

As with some DB's DDL is not transactional.  


> ... As a result of this migration cannot not be applied after first creation fails.

And what is the extension to MigrationRunner going to do?

Reply:
> Edit: I noticed that table creation should run in encapsulated transaction but the rollback does not work for table creation. Hence, tables are created, ini script fails, tables stills exists ==> endless loop.

Yes. In kubernetes the pods go into a (endless) crash loop [but that seems good and proper - to effectively fail the deployment].  

I suspect this leads to the question of whether migrations should have an associated "drop/rollback migration" (more like liquibase). There have been reasons for not doing this in the past. 

So yeah, I wonder where are you going with this?  Support for automatic "drop/rollback migration" ?

Reply:
Hi Rob,

1. Getter/Setter vs Protected Fields:
using a getter/setter pattern (also for usage inside the class) rather than using protected fields is more or less a question of code style. In the end, both is feasible. Pros & Cons are discussed multiple times. So, if you prefer protected fields, please go for it. My implementation is nothing more than a possible solution for the aspect to make the class a bit more flexible for others that have to find a solution for their own issue where MigrationRunner is one of many parts.

Sun recommends using getters in this special case, but it is a recommendation not a must-do. I personally declare fields final where ever I can and use Lombok to get rid of boiler plate code. So I used Lombok here as well and delomboked it to submit the PR.

2. DDL Migration
Unfortunately, I'm not very flexible here, as I just need to develop an extension for a complex, runtime-critical business application. The dump is over 1 GB in size. Hence, I can't afford to have the entire application go into stale mode. The application must support both SQL Server and Orcale. These are the requirements of the application vendor. At the moment the extension is working without ORM mapping. This is not the best way from my point of view. Since we had trouble in the past to get the manually created SQL statements in synch between Oracle and SQL Server.

That's how I came across EBEAN. Before I had implemented hibernate for testing purposes but the overhead is too big, too many dependencies I would have to shade into codebase. We are taking about 3 tables plus migration table at the moment. Most important is the reliability of the main application. So, unfortunately no containers, no AWS & Co., no microservices but legacy real life. :-)

If MigrationRunner would be more flexible for extending it I would be able to find a application driven approach for my use case. But I don't think, that it would be something that should be committed into your code base.

As I am new to ORM frameworks, when I looked at the _invoke method_ I assumed that a rollback is a rollback and not a maybe-rollback. As you mentioned it depends on the DB vendor that is used. Thanks for this information. I wasn't aware of!

Regards

Reply:
Right, cool.

> Getter/Setter vs Protected Fields

Yes I am very much on the protected field side. I believe we can just make migrationConfig protected, make run/invoke protected and remove both of the getter methods.   

Let me know if you want to update the PR to reflect that or if you want me to just make those changes.
 
Cheers, Rob.

Reply:
Sure. I will update the code. Thanks, Rob!